### PR TITLE
Fast evict leader

### LINKED ISSUE #3902: No need to wait tikv transfer leader timeout while upgrading tidb cluster with single tikv node
## Feature Request

**Is your feature request related to a problem? Please describe:**

It will take 10 minutes to wait tikv leader to transfer. Although there is only one tikv node. It's not necessary.

**Describe the feature you'd like:**

Upgrade the tikv directly if there is only one tikv

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by mianhk**
### What problem does this PR solve?
Close #3902 
Ref: https://github.com/pingcap/tidb-operator/pull/3918 

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [x] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Fast upgrade TiKV when tikv replicas number are less than 2
```

### CODE REVIEW

File: pkg/manager/member/tikv_upgrader.go [hunk 1]
```diff
@@ -149,6 +149,12 @@ func (u *tikvUpgrader) Upgrade(meta metav1.Object, oldSet *apps.StatefulSet, new
 			return nil
 		}
 
+		if len(podOrdinals) < 2 {
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_upgrader.go:152]: What about moving the check to L79 and just check the `*oldSet.Spec.Replicas`?

File: pkg/manager/member/tikv_upgrader.go [hunk 2]
```diff
@@ -203,6 +209,7 @@ func (u *tikvUpgrader) readyToUpgrade(upgradePod *corev1.Pod, tc *v1alpha1.TidbC
 	}
 
 	tlsEnabled := tc.IsTLSClusterEnabled()
+
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_upgrader.go:212]: No need to update this.

File: pkg/webhook/pod/tikv_deleter.go [hunk 3]
```diff
@@ -176,20 +176,26 @@ func (pc *PodAdmissionControl) admitDeleteUpTiKVPod(payload *admitPayload, store
...
+func (pc *PodAdmissionControl) admitDeleteUpTiKVPodDuringUpgrading(payload *admitPayload, store *pdapi.StoreInfo, stores *pdapi.StoresInfo) *admission.AdmissionResponse {
 	name := payload.pod.Name
 	namespace := payload.pod.Namespace
 	tc, ok := payload.controller.(*v1alpha1.TidbCluster)
 	if !ok {
 		err := fmt.Errorf("tikv pod[%s/%s]'s controller is not a tidbcluster", namespace, name)
 		return util.ARFail(err)
 	}
+
+	if tc.TiKVStsActualReplicas() < 2 {
```
Comment (DanielZhangQD) [inline pkg/webhook/pod/tikv_deleter.go:194]: Should use `TiKVStsDesiredReplicas`?

File: pkg/webhook/pod/tikv_deleter.go [hunk 4]
```diff
@@ -205,7 +211,7 @@ func (pc *PodAdmissionControl) admitDeleteUpTiKVPodDuringUpgrading(payload *admi
 		}
 	}
 
-	if !isTiKVReadyToUpgrade(payload.pod, store, tc.TiKVEvictLeaderTimeout()) {
+	if !isTiKVReadyToUpgrade(payload.pod, store, stores, tc.TiKVEvictLeaderTimeout()) {
```
Comment (DanielZhangQD) [inline pkg/webhook/pod/tikv_deleter.go:214]: No need to update this, right?

File: pkg/manager/member/tikv_upgrader.go (hunk 1 above)
Comment (mianhk) [reply pkg/manager/member/tikv_upgrader.go:152]: Ok, not clear for meam of the `partition` of statefulset before.

File: pkg/manager/member/tikv_upgrader.go [hunk 5]
```diff
@@ -77,6 +77,12 @@ func (u *tikvUpgrader) Upgrade(meta metav1.Object, oldSet *apps.StatefulSet, new
 		return fmt.Errorf("cluster[%s/%s] failed to upgrading tikv due to converting", meta.GetNamespace(), meta.GetName())
 	}
 
+	if *oldSet.Spec.Replicas < 2 {
+		klog.Infof("TiKV statefulset replicas are less than 2, skip waiting to evict region leader for tc %s/%s", ns, tcName)
+		setUpgradePartition(newSet, *oldSet.Spec.UpdateStrategy.RollingUpdate.Partition)
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_upgrader.go:82]:
```suggestion
		klog.Infof("TiKV statefulset replicas are less than 2, skip evicting region leader for tc %s/%s", ns, tcName)
		setUpgradePartition(newSet, 0)
```

File: pkg/manager/member/tikv_upgrader_test.go [hunk 6]
```diff
@@ -557,6 +557,42 @@ func TestTiKVUpgraderUpgrade(t *testing.T) {
...
+				tc.Status.TiKV.StatefulSet.Replicas = 1
+				tc.Status.TiKV.StatefulSet.CurrentReplicas = 1
+				tc.Status.TiKV.StatefulSet.UpdatedReplicas = 1
+			},
+			changeOldSet: func(oldSet *apps.StatefulSet) {
+				SetStatefulSetLastAppliedConfigAnnotation(oldSet)
+				oldSet.Status.CurrentReplicas = 1
+				oldSet.Status.UpdatedReplicas = 1
+				oldSet.Spec.Replicas = pointer.Int32Ptr(1)
+				oldSet.Spec.UpdateStrategy.RollingUpdate.Partition = pointer.Int32Ptr(0)
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_upgrader_test.go:575]:
```suggestion
				oldSet.Spec.UpdateStrategy.RollingUpdate.Partition = pointer.Int32Ptr(2)
```

File: pkg/webhook/pod/tikv_deleter.go [hunk 7]
```diff
@@ -190,6 +190,12 @@ func (pc *PodAdmissionControl) admitDeleteUpTiKVPodDuringUpgrading(payload *admi
 		err := fmt.Errorf("tikv pod[%s/%s]'s controller is not a tidbcluster", namespace, name)
 		return util.ARFail(err)
 	}
+
+	if tc.TiKVStsDesiredReplicas() < 2 {
+		klog.Infof("TiKV statefulset replicas are less than 2, skip waiting to evict region leader for Pod %s/%s", namespace, name)
```
Comment (DanielZhangQD) [inline pkg/webhook/pod/tikv_deleter.go:195]:
```suggestion
		klog.Infof("TiKV statefulset replicas are less than 2, skip evicting region leader for Pod %s/%s", namespace, name)
```

File: pkg/webhook/pod/tikv_util.go [hunk 8]
```diff
@@ -111,6 +111,7 @@ func isTiKVReadyToUpgrade(upgradePod *core.Pod, store *pdapi.StoreInfo, evictLea
 			klog.Errorf("parse annotation:[%s] to time failed.", EvictLeaderBeginTime)
 			return false
 		}
+
```
Comment (DanielZhangQD) [inline pkg/webhook/pod/tikv_util.go:114]: No need to update this.

**Comment by DanielZhangQD** [comment]
@mianhk Please add the test result in the PR description, thanks!

**Comment by mianhk** [comment]
> @mianhk Please add the test result in the PR description, thanks!

I add e2e test and fix promblems above. @DanielZhangQD PTAL.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 9]
```diff
@@ -327,6 +327,23 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+		tc.Spec.TiKV.Replicas = 1
+		// Deploy
+		utiltc.MustCreateTCWithComponentsReady(genericCli, oa, tc, 10*time.Minute, 5*time.Second)
+		ginkgo.By(fmt.Sprintf("Upgrading tidb cluster from %s to %s", tc.Spec.Version, utilimage.TiDBLatest))
+		err := controller.GuaranteedUpdate(genericCli, tc, func() error {
+			tc.Spec.Version = utilimage.TiDBLatest
+			return nil
+		})
+		framework.ExpectNoError(err, "failed to upgrade TidbCluster: %q", tc.Name)
+		err = oa.WaitForTidbClusterReady(tc, 10*time.Minute, 5*time.Second)
```
Comment (csuzhangxc) [inline tests/e2e/tidbcluster/tidbcluster.go:342]: this is hard to verify no evict leader is triggered 😂

**Review by csuzhangxc — APPROVED** [review]

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 10]
```diff
@@ -333,6 +333,31 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
 		framework.ExpectNoError(err, "failed to wait for TidbCluster ready: %q", tc.Name)
 	})
 
+	ginkgo.It("should direct upgrade tc successfully when TiKV replicas less than 2.", func() {
+		clusterName := "upgrade-cluster-tikv-1"
+		tc := fixture.GetTidbCluster(ns, clusterName, utilimage.TiDBLatestPrev)
+		tc.Spec.TiKV.Replicas = 1
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:339]:
```suggestion
		tc.Spec.TiKV.Replicas = 1
		tc.Spec.TiKV.EvictLeaderTimeout = pointer.StringPtr("10m")
```

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 9 above)
Comment (DanielZhangQD) [reply tests/e2e/tidbcluster/tidbcluster.go:348]:
```suggestion
		err := controller.GuaranteedUpdate(genericCli, tc, func() error {
			tc.Spec.Version = utilimage.TiDBLatest
			return nil
		})
		framework.ExpectNoError(err, "failed to upgrade TidbCluster: %q", tc.Name)
		err = oa.WaitForTidbClusterReady(tc, 5*time.Minute, 5*time.Second)
```

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**